frontend/src/components/pages/TranslationPage.tsx

```diff
@@ -0,0 +1,220 @@
+  ) {
+    updateStoryTranslationContentById(
+      storyTranslationContentData: {
+        id: $storyTranslationContentId
+        translationContent: $newTranslationContent
+      }
+    ) {
+      story {
+        lineIndex
+      }
+    }
+  }
+`;
+
+const TranslationPage = () => {
+  const {
+    storyIdParam,
+    storyTranslationIdParam,
+  } = useParams<TranslationPageProps>();
+
+  // do we need to catch these?
+  const storyId = +storyIdParam!!;
+  const storyTranslationId = +storyTranslationIdParam!!;
+
+  const [translatedStoryLines, setTranslatedStoryLines] = useState<StoryLine[]>(
+    [],
+  );
+
+  const handleError = (errorMessage: string) => {
+    alert(errorMessage);
+  };
+
+  const arrayIndex = (lineIndex: number): number =>
+    lineIndex - (translatedStoryLines[0].lineIndex - 1);
+
+  // TODO replace with real logic
+  const addIcon = (): JSX.Element | null => {
+    const randomNum = Math.floor(Math.random() * 3);
+
+    switch (randomNum) {
+      case 0:
+        return <img src={CheckmarkIcon} alt="Approved line" />;
+      case 1:
+        return <img src={CommentIcon} alt="Line with feedback" />;
+      default:
+        return null;
+    }
+  };
+
+  const [updateTranslation] = useMutation<{
+    updateStoryTranslationContentById: UpdateTranslationResponse;
+  }>(UPDATE_TRANSLATION);
+  const onChangeTranslationContent = async (
+    storyTranslationContentId: number,
+    newContent: string,
+  ) => {
+    try {
+      const result = await updateTranslation({
+        variables: {
+          storyTranslationContentId,
+          newTranslationContent: newContent,
+        },
+      });
+
+      const lineIndex =
+        result.data?.updateStoryTranslationContentById.story.lineIndex;
+      if (lineIndex !== undefined) {
+        // this is update only, so id already set
+        // not convinced this is the best strateg, this will re-redner the page
+        // and the box that has already been updated right?
+        // should only update whatever is storingi nfo without re-rendering if possible
+        const updatedContentArray = [...translatedStoryLines];
+        const index = arrayIndex(lineIndex);
+        updatedContentArray[index].translatedContent = newContent;
+
+        setTranslatedStoryLines(updatedContentArray);
+      } else {
+        handleError("Unable to save translation.");
+      }
+    } catch (err) {
+      handleError(err ?? "Error occurred, please try again.");
+    }
+  };
+
+  useQuery(GET_STORY_CONTENTS(storyId, storyTranslationId), {
+    fetchPolicy: "cache-and-network",
+    onCompleted: (data) => {
+      const storyContent = data.storyById.contents;
+      const translatedContent = data.storyTranslationById.translationContents;
+
+      const contentArray: StoryLine[] = [];
+      // would callign directly onto state object rather than this array re-render
+      // expect an ordered response
+      storyContent.forEach(({ id, content, lineIndex }: Content) => {
```

Comment:
Proper order on the page assumes that `storyContent` will be ordered by lineIndex, which I'm not sure if I can guarantee unless that SQL query has a nice ORDER BY default. Should the array (since it's mapped) be sorted in the frontend or place an ORDER BY on the query endpoint?

Reply:
Also, just double-checking - would altering the state object `translatedStoryLines` instead of using a local const `contentArray` force a re-render each iteration of `forEach`? I assumed it would but wasn't sure

Reply:
Not sure how zipping would look. Without zipping, I'm thinking the options would be
- 2 sorts (both response arrays) + one iteration (go through both together by index)
- 1 sort (first response array) + 2 iterations (go through each separately)

I think second should be more efficient?

Reply:
^ response on below comment: https://github.com/uwblueprint/planet-read/pull/39/files#r667644364